docs/docs/programmatically-create-pages-from-data.md

```diff
@@ -2,7 +2,108 @@
 title: Programmatically create pages from data
 ---
 
-This is a stub. Help our community expand it.
+Gatsby and its ecosystem of plugins provide all kinds of data through a
+GraphQL interface. This guide will show how that data can be used to
+programmatically create pages.
 
-Please use the [Gatsby Style Guide](/docs/gatsby-style-guide/) to ensure your
-pull request gets accepted.
+### Prerequisites
+
+Though you can use any data source you'd like, this guide will show how to
+create pages from markdown files (following after the example introduced in
+[earlier guides](https://www.gatsbyjs.org/docs/adding-markdown-pages/)).
+
+### Creating Pages
+
+The Gatsby Node API provides the
+[`createPages`](https://www.gatsbyjs.org/docs/node-apis/#createPages)
+extension point which we'll use to add pages. This function will give us
+access to the
+[`createPage`](https://www.gatsbyjs.org/docs/actions/#createPage) action
+which is at the core of programmatically creating a page.
+
+```javascript{17-25}:title=gatsby-node.js
+exports.createPages = ({ graphql, actions }) => {
+  const { createPage } = actions;
+  return new Promise((resolve, reject) => {
+    graphql(`
+      {
+        allMarkdownRemark {
+          edges {
+            node {
+              fields {
+                slug
+              }
+            }
+          }
+        }
+      }
+    `).then(result => {
+      result.data.allMarkdownRemark.edges.forEach(({ node }) => {
+        createPage({
+          path: node.fields.slug,
+          component: path.resolve(`./src/templates/blog-post.js`),
+          context: {
+            slug: node.fields.slug,
+          },
+        });
+      });
+      resolve();
+    });
+  });
+};
+```
+
+For each page we want to create we must specify the `path` for visiting that
+page, the `component` template used to render that page, and any `context`
```

Comment:
maybe mention that `context` is optional? By default gatsby will pass some context already - stuff like `path`